spec/syntax.md

```diff
@@ -0,0 +1,598 @@
+
+### Message
+
+A single message is either a plain message, a single pattern, or has a preamble
+followed by one or more variants which represent the translatable body of the message.
+
+```ebnf
+Message ::= Plain | Pattern | Preamble Variant+
+```
+
+### Plain
+
+A plain message only contains translatable content;
+placeholders or their delimiters are not allowed inside a plain message.
+Plain messages must not start with one of the syntax characters `[`, `{` or `$`,
+as those would indicate that the message has a more complex structure.
+Any whitespace at the beginning or end of a plain message is ignored.
+A plain message cannot represent an empty string;
+for that, use an empty pattern `[]` instead.
+
+```ebnf
+Plain ::= PlainStart (PlainChar* PlainEnd)?  /* ws: explicit */
+PlainChar ::= AnyChar - ('{' | '}')
+PlainStart ::= PlainChar - ('[' | '$' | WhiteSpace)
+PlainEnd ::= PlainChar - WhiteSpace
+```
+
+### Preamble
+
+The preamble is where selectors and local variables can be defined.
+A selector is an expression which will be used to choose one of the variants during formatting.
+A selector can be optionally bound to a local variable, which may then be used in other expressions.
+
+```ebnf
+Preamble ::= Selector+
+Selector ::= (Variable '=')? '{' Expression '}'
+```
+
+Examples:
+
+```
+$frac = {$count: number minFractionDigits=2}
+    one [One apple]
+    _ [{$frac} apples]
+```
+
+### Variants
+
+A variant is a keyed pattern.
+The keys are used to match against the selectors defined in the preamble.
+
+```ebnf
+Variant ::= VariantKey* Pattern
+VariantKey ::= String | Nmtoken
+```
+
+A well-formed message is considered valid if the following requirements are satisfied:
+
+- The number of keys on each variant must be fewer or equal to the number of selectors defined in the preamble.
+- At least one variant's keys must all be equal to the catch-all key (`_`).
+
+### Patterns
+
+A pattern is a sequence of translatable elements.
+Patterns are always delimited with `[` at the start, and `]` at the end.
+This serves 3 purposes:
+
+- The message should be unambiguously embeddable in various container formats
+  regardless of the container's whitespace trimming rules.
+  E.g. in Java `.properties` files,
+  `hello = [Hello]` will unambiguously define the `Hello` message without the space in front of it.
+- The message should be conveniently embeddable in various programming languages
+  without the need to escape characters commonly related to strings, e.g. `"` and `'`.
+  Such need may still occur when a singe or double quote is
+  used in the translatable content or to delimit a string literal.
+- The syntax should make it as clear as possible which parts of the message body
+  are translatable and which ones are part of the formatting logic definition.
+
+```ebnf
+Pattern ::= '[' (Text | Placeable)* ']' /* ws: explicit */
+```
+
+Examples:
+
+```
+[Hello, world!]
```

Comment:
The current syntax (intentionally) allows for simple messages without `[`, `]` (no selectors, no placeholders)
So this is not accurate.

In fact, the current syntax allows both `Hello world!` and `[Hello world!]`, which I don't think is a feature.

But for this PR I'm not arguing that, only that the examples here should reflect the current grammar.

Reply:
This section is describing the `Pattern` construction, which does require `[` `]` as delimiters, and for which `[Hello, world!]` is a valid example. The un-delimited form is called `Plain`, and is described between the `Message` and `Preamble` sections above.

Reply:
So both `[Hello, world!]` and `Hello, world!` are valid?
And worse, parsing them result in different representations in the data model?
But give the same result when rendered?

If that is intentional, let me know and I will open an issue, because that looks like bad design in my opinion.

Reply:
Yes, it is intentional that both `[Hello, world!]` and `Hello, world!` are valid but different representations of a message that would format to a string as `Hello, world!`.

This is comparable to how some languages and formats allow for both `'Hello, world!'` and `"Hello, world!"` to be representations of the same string, while having different handling for their contents (example: shell scripts).

Reply:
`Pain` and `Pattern` are different grammar productions, but they both result in the same data model.

Reply:
The " vs ' in some programming languages don't do the same thing:
```
# export name=Mihai
# echo 'Hello $name'
Hello $name
# echo "Hello $name"
Hello Mihai
```

Hawing several ways to do the same things is gratuitous.

TMTOWTDI is not currently seen as a good thing
https://en.wikipedia.org/wiki/There%27s_more_than_one_way_to_do_it

Reply:
> TMTOWTDI is not currently seen as a good thing

I don't think this is a good analogy. The wikipedia article you provided ends with a quote from Python Zen, and Python is exactly allowing for multiple ways to format a string, with different tradeoffs: https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals

The current proposal uses the tradeoff between those two modes.

Reply:
Let's continue this discussion in a separate issue dedicated to the topic?